src/bloqade/noise/fidelity.py

```diff
@@ -0,0 +1,43 @@
+from kirin import interp
+from kirin.lattice import EmptyLattice
+
+from bloqade.analysis.fidelity import FidelityAnalysis
+
+from .native import dialect as native
+from .native.stmts import PauliChannel, CZPauliChannel, AtomLossChannel
+
+
+@native.register(key="circuit.fidelity")
+class FidelityMethodTable(interp.MethodTable):
+
+    @interp.impl(PauliChannel)
+    @interp.impl(CZPauliChannel)
+    def pauli_channel(
+        self,
+        interp: FidelityAnalysis,
+        frame: interp.Frame[EmptyLattice],
+        stmt: PauliChannel | CZPauliChannel,
+    ):
+        probs = stmt.probabilities
+        try:
+            ps, ps_ctrl = probs
+        except ValueError:
+            (ps,) = probs
+            ps_ctrl = ()
+
+        p = sum(ps)
+        p_ctrl = sum(ps_ctrl)
+
+        # NOTE: fidelity is just the inverse probability of any noise to occur
+        fid = (1 - p) * (1 - p_ctrl)
+
+        interp._current_gate_fidelity *= fid
+
+    @interp.impl(AtomLossChannel)
+    def atom_loss(
+        self,
+        interp: FidelityAnalysis,
+        frame: interp.Frame[EmptyLattice],
+        stmt: AtomLossChannel,
+    ):
+        interp._current_atom_survival_probability *= 1 - stmt.prob
```

Comment:
Would it be difficult to split this up to track the loss probability of individual qubits?

Reply:
Not sure. How would you keep track of which qubit is which? It would work for `PyQrackQubit`s because I can use the `addr`, but this is more general, and I just have an `IList` of qubits.

So you could have something like this:
```python
AtomLossChannel(prob=0.001, qargs=[q[0], q[1]])
AtomLossChannel(prob=0.001, qargs=[q[0], q[2]])
```
Do you have an idea how I could tell that we have `q[2]` in the second entry in the second line?

Reply:
There are two options I guess: 

1. Create a new lattice and have a lattice element that accounts for the qubit values
2. Reuse the AddressAnalysis results and pass that into this interpreter and use those values inside the interpreter to map the qubit ssa values to qubit ids.